packages/gator-permissions-snap/src/core/rules.tsx

```diff
@@ -156,18 +207,107 @@ export function bindRuleHandlers<
       );
       await onContextChanged({ context: updatedContext });
     };
-    userEventDispatcher.on({
-      elementName: name,
-      eventType: UserInputEventType.InputChangeEvent,
-      interfaceId,
-      handler: handleInputChange,
-    });
-
-    acc.push({
-      elementName: name,
-      eventType: UserInputEventType.InputChangeEvent,
-      handler: handleInputChange as UserEventHandler<UserInputEventType>,
-    });
+
+    const handleDateInputChange: UserEventHandler<
+      UserInputEventType.InputChangeEvent
+    > = async ({ event }) => {
+      const fieldName = event.name;
+      const isDateField = fieldName.endsWith('_date');
+      const isTimeField = fieldName.endsWith('_time');
+
+      // Fix race condition: Get context once and reuse it
```

Comment:
nit: There's a couple comments which, while they might be helpful, have strange tense.

```
// Get context once and reuse it to avoid race condition
```